Add support for DA monitor binary station data files

EdwardSafford-NOAA opened the pull request:
## Description

Extend `data/mon_data_space.py` to support legacy DA monitor binary station data files.  These are a binary files with a header on each observation record. The header includes the station data (station id, lat, lon, etc). The accompanying control file only indicates the variable information but does not include the number of observations.  This can only be learned by reading the data file.

This completes issue #84 .

## Dependencies

None

## Impact

None

EdwardSafford-NOAA committed:
Ref #84
save work in progress

EdwardSafford-NOAA committed:
Ref #84

Save work in progress.

EdwardSafford-NOAA committed:
Ref #84

Save work in progress

EdwardSafford-NOAA committed:
Ref #84
Save work in progress

EdwardSafford-NOAA committed:
Ref #84

Merge branch 'develop' into feature/stn-84

EdwardSafford-NOAA committed:
Ref #84

Clean up pycode issues

CoryMartin-NOAA commented on src/eva/data/mon_data_space.py:
```diff
@@ -77,43 +92,70 @@ def execute(self, dataset_config, data_collections, timing):
+                                                                    ndims_used, dims_arr, nvars,
+                                                                    stn_vars)
+                y_range = np.arange(1, dims['ydef']+1)
+                coords['ydef'] = "Nobs" + str(file_ctr)
```
Does this string need a space in it? `"Nobs " + str(file_ctr)`?

EdwardSafford-NOAA replied:
I guess it could, but I don't think it's necessary.  The problem is that the number of obs is file dependent.  So I need a means to associate the variables in a given file with the correct coordinates and dimensions when loading more than one (the ozn horiz plots load 4 files).  This solution gives me variables obs1, ges1, etc associated with dimension nobs1 and coordinate Nobs1, and then a different set with the next file.  Horiz plots typically load 1-4 files.

Alternatively I could figure out the greatest number of obs in the fileset and pad the shorter ones out with NaN values so the dimensions could be shared by all variables. That would be a bit more of more effort in the data load but I'll be glad to go that way if you'd prefer.  Now that I describe it I think I like it better than what I've got now.

CoryMartin-NOAA replied:
Ok I understand better what the string concatenation is doing. I'll defer to you. If you are happy with what you have, so am I but I also will support your suggestion of shared dimensions (it would probably be cleaner)

EdwardSafford-NOAA replied:
Let me give that a try.  If it doesn't take to long it's probably worth it.  I'll switch this PR to draft.

EdwardSafford-NOAA replied:
Revising that was the right answer.  I learned about xarray.align() and the awesome power that wields.  One call syncs the dimensions of all datasets and fills missing values with NaN.  It sure is easy when you know how. :)

CoryMartin-NOAA commented on src/eva/data/mon_data_space.py:
```diff
@@ -77,43 +92,70 @@ def execute(self, dataset_config, data_collections, timing):
                 timestep_ds.attrs['sensor'] = attribs['sensor']
 
+            if len(lat):
+                lat_str = 'lat' + str(file_ctr) if stn_data else 'lat'
```
Similar comments on these lines

CoryMartin-NOAA reviewed (commented):
Couple brief questions/comments but otherwise looks good from what I can tell

EdwardSafford-NOAA committed:
Ref #84

Align station data datasets, filing in NaN for all missing values.
Merge as per non-station datasets and we're golden.

CoryMartin-NOAA reviewed (approved):
Looks good, thanks!
